### PR TITLE
[pme] Added save and restore functionality

### PR DESCRIPTION
**Pull request by jimmy-huang**

### CODE REVIEW

**Comment by jimmy-huang** [comment]
@poussa please review

File: arc/src/main.c [hunk 1]
```diff
@@ -1051,27 +1061,29 @@ static void handle_pme(struct zjs_ipm_message* msg)
...
-                                                                    data.conext & NCR_CONTEXT,
-                                                                    data.aif,
-                                                                    data.min_if
-                                                                    data.category);
+        DBG_PRINT("neuron: id=%d, CTX=%d, AIF=%d MIF=%d, cat=%d\n",
+                  data.context & NCR_ID,
+                  data.context & NCR_CONTEXT,
+                  data.aif,
+                  data.min_if,
+                  data.category);
```
Comment (grgustaf) [inline arc/src/main.c]: I'd pack these as tight as possible in 80 chars, e.g. use 2-3 lines. Especially for a debug print, there's no need for special clarity in arguments - it's not real code you care about reading carefully, so it's just taking up more screen space.

File: arc/src/main.c [hunk 2]
```diff
@@ -1089,20 +1101,25 @@ static void handle_pme(struct zjs_ipm_message* msg)
...
         msg->data.pme.g_context = curie_pme_get_global_context();
-        DBG_PRINT("global context: %d\n", msg->data.pme.g_context);
+        DBG_PRINT("global context: %d\n",
+                  msg->data.pme.g_context);
         break;
     case TYPE_PME_SET_GLOBAL_CONTEXT:
         // valid range is 1-127
-        if (msg->data.pme.g_context < 1 || msg->data.pme.g_context > 127) {
+        if (msg->data.pme.g_context < 1 ||
+            msg->data.pme.g_context > 127) {
```
Comment (grgustaf) [inline arc/src/main.c]: I really don't think this is more readable either, multi-line if conditions are best avoided...

File: arc/src/main.c [hunk 3]
```diff
@@ -1089,20 +1101,25 @@ static void handle_pme(struct zjs_ipm_message* msg)
...
         msg->data.pme.committed_count = curie_pme_get_committed_count();
-        DBG_PRINT("committed count: %d\n", msg->data.pme.committed_count);
+        DBG_PRINT("committed count: %d\n",
+                  msg->data.pme.committed_count);
         break;
     case TYPE_PME_GET_GLOBAL_CONTEXT:
         msg->data.pme.g_context = curie_pme_get_global_context();
-        DBG_PRINT("global context: %d\n", msg->data.pme.g_context);
+        DBG_PRINT("global context: %d\n",
+                  msg->data.pme.g_context);
```
Comment (grgustaf) [inline arc/src/main.c]: same with these definitely

File: src/zjs_pme.h [hunk 4]
```diff
@@ -15,4 +15,4 @@ jerry_value_t zjs_pme_init();
 /** Release resources held by the pme module */
 void zjs_pme_cleanup();
 
-#endif  // __zjs_pme_h__
+#endif // __zjs_pme_h__
```
Comment (grgustaf) [inline src/zjs_pme.h]: I purposely put two spaces before any same-line comment FWIW.

File: src/zjs_pme.c [hunk 5]
```diff
@@ -186,29 +200,32 @@ static ZJS_DECL_FUNC(zjs_pme_read_neuron)
...
+
+    memset(send.data.pme.vector, 0, sizeof(uint8_t) * MAX_VECTOR_SIZE);
+    for (int i = 0; i < vector_len; i++) {
+        ZVAL num = jerry_get_property_by_index(vector, i);
+        if (!jerry_value_is_number(num)) {
             return zjs_error("invalid vector type");
         }
-        uint8_t byte = jerry_get_number_value(vector);
+        uint8_t byte = jerry_get_number_value(num);
         memcpy(&send.data.pme.vector[i], &byte, sizeof(uint8_t));
```
Comment (grgustaf) [inline src/zjs_pme.c]: This is aka `send.data.pme.vector[i] = byte;` which is preferable. :)

File: src/zjs_pme.c [hunk 6]
```diff
@@ -295,13 +315,142 @@ static ZJS_DECL_FUNC(zjs_pme_get_distance_mode)
...
+        send.data.pme.min_if = min_if;
+
+        memset(send.data.pme.vector, 0, sizeof(uint8_t) * MAX_VECTOR_SIZE);
+        for (int i = 0; i < vector_len; i++) {
+            ZVAL num = jerry_get_property_by_index(vector, i);
+            if (!jerry_value_is_number(num)) {
+                 return zjs_error("invalid vector type");
+            }
+            uint8_t byte = jerry_get_number_value(num);
+            memcpy(&send.data.pme.vector[i], &byte, sizeof(uint8_t));
```
Comment (grgustaf) [inline src/zjs_pme.c]: Don't use memcpy for this, again.

**Comment by grgustaf** [comment]
+1

**Pull request merged**